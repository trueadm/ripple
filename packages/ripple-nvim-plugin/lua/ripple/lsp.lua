local M = {}

local SERVER_NAME = "ripple"
local LSP_PACKAGE = "ripple-language-server"
local EXACT_VERSION_PATTERN = "^%d+%.%d+%.%d+$"

local function plugin_package_json_path()
	local plugin_root = debug.getinfo(1, "S").source:sub(2)
	return vim.fs.find("package.json", { upward = true, path = vim.fs.dirname(plugin_root) })[1]
end

local function read_json(path)
	local stat = vim.loop.fs_stat(path)
	if not stat or stat.type ~= "file" then
		return nil
	end

	local ok, contents = pcall(vim.fn.readfile, path)
	if not ok or type(contents) ~= "table" then
		return nil
	end

	local json_string = table.concat(contents, "")

	local decoded
	local ok_decode, value = pcall(vim.json.decode, json_string)
	if ok_decode then
		decoded = value
	end

	if type(decoded) ~= "table" then
		return nil
	end

	return decoded
end

local function resolve_required_version()
	local package_json_path = plugin_package_json_path()

	local package_json = read_json(package_json_path)
	if not package_json then
		return nil, "unable to read plugin package.json"
	end

	local config = package_json.config or {}

	local spec = config["ripple-language-server"]

	if type(spec) ~= "string" or spec == "" then
		return nil, "missing config.rippleLanguageServerVersion field in package.json"
	end

	spec = spec:gsub("^%s+", ""):gsub("%s+$", "")

	if not spec:match(EXACT_VERSION_PATTERN) then
		return nil, ("unsupported version spec '%s'; please pin an exact semver"):format(spec)
	end

	return spec, nil
end

local function installed_version(install_dir)
	local pkg = read_json(install_dir .. "/node_modules/" .. LSP_PACKAGE .. "/package.json")
	if not pkg or type(pkg.version) ~= "string" then
		return nil
	end
	return pkg.version
end

local function ensure_server_binary()
	local required_version, err = resolve_required_version()
	if not required_version then
		vim.notify("[ripple] " .. err, vim.log.levels.ERROR)
		return nil
	end

	local install_dir = vim.fn.stdpath("data") .. "/" .. LSP_PACKAGE
	local bin = install_dir .. "/node_modules/.bin/" .. LSP_PACKAGE

	if vim.loop.os_uname().version:match("Windows") then
		bin = bin .. ".cmd"
	end

	local function version_matches()
		local version = installed_version(install_dir)
		return version == required_version
	end

	if vim.fn.executable(bin) == 1 and version_matches() then
		return bin
	end

	vim.fn.mkdir(install_dir, "p")
	vim.notify(("[ripple] Installing %s@%s..."):format(LSP_PACKAGE, required_version), vim.log.levels.INFO)

	local out = vim.fn.system({
		"npm",
		"install",
		("%s@%s"):format(LSP_PACKAGE, required_version),
		"--prefix",
		install_dir,
		"--no-audit",
		"--no-fund",
	})

	if vim.v.shell_error ~= 0 then
		vim.notify("[ripple] npm install failed:\n" .. out, vim.log.levels.ERROR)
		return nil
	end

	if not version_matches() then
		local found = installed_version(install_dir) or "unknown"
		vim.notify(
			("[ripple] Installed %s but required %s"):format(found, required_version),
			vim.log.levels.ERROR
		)
		return nil
	end

	return bin
end

function M.setup()
	local bin = ensure_server_binary()
	if not bin then
		return
	end

	local base_config = {
		cmd = { bin, "--stdio" },
		filetypes = { "ripple" },
		root_markers = { "package.json", "pnpm-workspace.yaml", ".git" },
	}

	vim.lsp.config(SERVER_NAME, base_config)
	vim.lsp.enable(SERVER_NAME)
end

return M
