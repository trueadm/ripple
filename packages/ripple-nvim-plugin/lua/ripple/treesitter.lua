local M = {}

function M.setup()
	local ok, parsers = pcall(require, "nvim-treesitter.parsers")
	if not ok then
		vim.schedule(function()
			vim.notify("[ripple] nvim-treesitter not found; skipping parser registration", vim.log.levels.WARN)
		end)
		return
	end

	local parser_config = parsers.get_parser_configs()
	local default_install = {
		url = "https://github.com/trueadm/ripple",
		files = {
			"src/parser.c",
			"src/scanner.c",
		},
		branch = "main",
		location = "packages/tree-sitter-ripple",
	}

	local config = parser_config.ripple or {}
	local install_info = config.install_info or {}
	config.install_info = vim.tbl_deep_extend("force", default_install, install_info)
	config.filetype = config.filetype or "ripple"
	config.used_by = config.used_by or { "ripple" }
	parser_config.ripple = config

	local ok_install, install = pcall(require, "nvim-treesitter.install")
	if ok_install and type(install.ensure_installed) == "function" then
		install.ensure_installed("ripple")
	else
		vim.schedule(function()
			vim.notify("[ripple] Install nvim-treesitter parser manually via :TSInstall ripple", vim.log.levels.INFO)
		end)
	end
end

return M
