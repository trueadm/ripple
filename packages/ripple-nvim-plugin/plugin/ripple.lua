if vim.filetype and vim.filetype.add then
	vim.filetype.add({
		extension = {
			ripple = "ripple",
		},
	})
end
