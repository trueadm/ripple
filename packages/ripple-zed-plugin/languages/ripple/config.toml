name = "Ripple"
grammar = "ripple"
path_suffixes = ["ripple"]
line_comments = ["//"]
block_comment = ["/*", "*/"]
tab_size = 2
hard_tabs = false

autoclose_before = ";:.,=}])\""

brackets = [
  { start = "{", end = "}", close = true, newline = true },
  { start = "[", end = "]", close = true, newline = true },
  { start = "(", end = ")", close = true, newline = true },
  { start = "<", end = ">", close = true, newline = false },
  { start = "\"", end = "\"", close = true, newline = false },
  { start = "'", end = "'", close = true, newline = false },
  { start = "`", end = "`", close = true, newline = false },
]
