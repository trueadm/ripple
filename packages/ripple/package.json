{
  "name": "ripple",
  "description": "Ripple is an elegant TypeScript UI framework",
  "license": "MIT",
  "author": "Dominic Gannaway",
  "version": "0.2.67",
  "type": "module",
  "module": "src/runtime/index.js",
  "main": "src/runtime/index.js",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/trueadm/ripple.git",
    "directory": "packages/ripple"
  },
  "bugs": {
    "url": "https://github.com/trueadm/ripple/issues"
  },
  "homepage": "https://ripplejs.com",
  "keywords": [
    "ripple",
    "UI",
    "framework"
  ],
  "exports": {
    ".": {
      "types": "./types/index.d.ts",
      "worker": "./src/runtime/index.js",
      "browser": "./src/runtime/index.js",
      "default": "./src/runtime/index.js"
    },
    "./package.json": "./package.json",
    "./compiler": {
      "types": "./types/index.d.ts",
      "require": "./compiler/index.js",
      "default": "./src/compiler/index.js"
    },
    "./validator": {
      "types": "./types/index.d.ts",
      "require": "./validator/index.js",
      "default": "./src/validator/index.js"
    },
    "./internal/client": {
      "default": "./src/runtime/internal/client/index.js"
    },
    "./jsx-runtime": {
      "types": "./src/jsx-runtime.d.ts",
      "import": "./src/jsx-runtime.js",
      "default": "./src/jsx-runtime.js"
    }
  },
  "imports": {
    "#client": "./src/runtime/internal/client/types.d.ts",
    "#client/constants": "./src/internal/client/constants.js"
  },
  "dependencies": {
    "@jridgewell/sourcemap-codec": "^1.5.5",
    "acorn": "^8.15.0",
    "acorn-typescript": "^1.4.13",
    "esrap": "^2.1.0",
    "is-reference": "^3.0.3",
    "magic-string": "^0.30.18",
    "muggle-string": "^0.4.1",
    "zimmerframe": "^1.1.2"
  },
  "devDependencies": {
    "@types/estree": "^1.0.8",
    "@types/node": "^24.3.0",
    "typescript": "catalog:"
  }
}
