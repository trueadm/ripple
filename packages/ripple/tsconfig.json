{
	"compilerOptions": {
		"module": "esnext",
		"lib": ["esnext", "dom", "dom.iterable"],
		"target": "esnext",
		"noEmit": true,
		"moduleResolution": "bundler",
		"resolveJsonModule": true,
		"noEmitOnError": true,
		"noErrorTruncation": true,
		"allowSyntheticDefaultImports": true,
		"verbatimModuleSyntax": true,
		"types": ["node"],
		"jsx": "preserve",
		"jsxImportSource": "ripple",
		"strict": true,
		"allowJs": true,
		"checkJs": true,
		"paths": {
			"ripple": ["./types/index.d.ts"],
		}
	},
	"include": [
		"./*.js",
		"./src/",
		"./tests/*/test.ripple",
	]
}
