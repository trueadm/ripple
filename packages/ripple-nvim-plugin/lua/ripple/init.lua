local M = {}

---Entrypoint invoked by plugin managers.
---@param opts? {treesitter?: table|boolean, lsp?: table|boolean}
function M.setup(opts)
	opts = opts or {treesitter = { ensure_installed = true }, lsp = {}}

	if opts.treesitter ~= false then
		require("ripple.treesitter").setup(opts.treesitter or {})
	end

	if opts.lsp ~= false then
		require("ripple.lsp").setup(opts.lsp or {})
	end
end

return M
