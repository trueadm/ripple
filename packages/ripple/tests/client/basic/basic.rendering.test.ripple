import { describe, it, expect } from 'vitest';
import { track, flushSync } from 'ripple';

describe('basic client > rendering & text', () => {
	it('renders static text', () => {
		component Basic() {
			<div>{'Hello World'}</div>
		}

		expect(container).toBeDefined();

		render(Basic);
		expect(container).toMatchSnapshot();
	});

	it('renders semi-dynamic text', () => {
		component Basic() {
			let text = 'Hello World';

			<div>{text}</div>
		}

		render(Basic);

		expect(container).toMatchSnapshot();
	});

	it('renders dynamic text', () => {
		component Basic() {
			let text = track('Hello World');

			<button onClick={() => { @text = 'Hello Ripple' }}>{'Change Text'}</button>
			<div>{@text}</div>
		}

		render(Basic);

		const button = container.querySelector('button');

		button.click();
		flushSync();

		expect(container.querySelector('div').textContent).toEqual('Hello Ripple');
	});

	it('renders empty string literal', () => {
		component Basic() {
			<div>{''}</div>
		}

		render(Basic);
		expect(container.querySelector('div').textContent).toEqual('');
	});

	it('renders empty template literal', () => {
		component Basic() {
			<div>{``}</div>
		}

		render(Basic);
		expect(container.querySelector('div').textContent).toEqual('');
	});

	it('renders tick template literal for nested children', () => {
		component Child({ level, children }: { level: number, children: any }) {
			if(level == 1) {
				<h1><children /></h1>
			}
			if(level == 2) {
				<h2><children /></h2>
			}
			if(level == 3) {
				<h3><children /></h3>
			}
		}

		component App() {
			<Child level={1}>{`Heading 1`}</Child>
		}

		render(App);
		expect(container.querySelector('h1').textContent).toEqual('Heading 1');
	});

	it('renders simple JS expression logic correctly', () => {
		component Example() {
			let test = {}
			let counter = 0;
			test[counter++] = 'Test';

			<div>{JSON.stringify(test)}</div>
			<div>{JSON.stringify(counter)}</div>
		}
		render(Example);

		expect(container).toMatchSnapshot();
	});

	it('renders with mixed static and dynamic content', () => {
		component Basic() {
			let name = track('World');
			let count = track(0);
			const staticMessage = 'Welcome to Ripple!';

			<div class='mixed-content'>
				<h1>{staticMessage}</h1>
				<p class='greeting'>{'Hello, ' + @name + '!'}</p>
				<p class='notifications'>{'You have ' + @count + ' notifications'}</p>
				<button onClick={() => { @count++ }}>{'Add Notification'}</button>
				<button onClick={() => { @name = @name === 'World' ? 'User' : 'World' }}>{'Toggle Name'}</button>
			</div>
		}

		render(Basic);

		const heading = container.querySelector('h1');
		const greetingP = container.querySelector('.greeting');
		const notificationsP = container.querySelector('.notifications');
		const buttons = container.querySelectorAll('button');

		expect(heading.textContent).toBe('Welcome to Ripple!');
		expect(greetingP.textContent).toBe('Hello, World!');
		expect(notificationsP.textContent).toBe('You have 0 notifications');

		buttons[0].click();
		flushSync();
		expect(notificationsP.textContent).toBe('You have 1 notifications');

		buttons[1].click();
		flushSync();
		expect(greetingP.textContent).toBe('Hello, User!');
	});

	it('basic operations', () => {
		component App() {
			let count = track(0)
			<div>{@count++}</div>
			<div>{++@count}</div>
			<div>{5}</div>
			<div>{@count}</div>
		}

		render(App);
		expect(container).toMatchSnapshot();
	});

	it('renders with conditional rendering using if statements', () => {
		component Basic() {
			let showContent = track(false);
			let userRole = track('guest');

			<button onClick={() => { @showContent = !@showContent }}>{'Toggle Content'}</button>
			<button onClick={() => { @userRole = @userRole === 'guest' ? 'admin' : 'guest' }}>{'Toggle Role'}</button>

			<div class='content'>
				if (@showContent) {
					if (@userRole === 'admin') {
						<div class='admin-content'>{'Admin content'}</div>
					} else {
						<div class='user-content'>{'User content'}</div>
					}
				} else {
					<div class='no-content'>{'No content'}</div>
				}
			</div>
		}

		render(Basic);

		const buttons = container.querySelectorAll('button');
		const contentDiv = container.querySelector('.content');

		expect(contentDiv.querySelector('.no-content')).toBeTruthy();
		expect(contentDiv.querySelector('.admin-content')).toBeFalsy();
		expect(contentDiv.querySelector('.user-content')).toBeFalsy();

		buttons[0].click();
		flushSync();

		expect(contentDiv.querySelector('.no-content')).toBeFalsy();
		expect(contentDiv.querySelector('.user-content')).toBeTruthy();
		expect(contentDiv.querySelector('.admin-content')).toBeFalsy();

		buttons[1].click();
		flushSync();

		expect(contentDiv.querySelector('.no-content')).toBeFalsy();
		expect(contentDiv.querySelector('.user-content')).toBeFalsy();
		expect(contentDiv.querySelector('.admin-content')).toBeTruthy();
	});

	it('should handle lexical scopes correctly', () => {
		component App() {
			<section>
				let sectionData = 'Nested scope variable';

				{sectionData}
			</section>
		}

		render(App);
		expect(container).toMatchSnapshot();
	});
});
