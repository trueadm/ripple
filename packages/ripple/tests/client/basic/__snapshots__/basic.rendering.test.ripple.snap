// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic client > rendering & text > basic operations 1`] = `
<div>
  <div>
    0
  </div>
  <div>
    2
  </div>
  <div>
    5
  </div>
  <div>
    2
  </div>
  
</div>
`;

exports[`basic client > rendering & text > renders semi-dynamic text 1`] = `
<div>
  <div>
    Hello World
  </div>
  
</div>
`;

exports[`basic client > rendering & text > renders simple JS expression logic correctly 1`] = `
<div>
  <div>
    {"0":"Test"}
  </div>
  <div>
    1
  </div>
  
</div>
`;

exports[`basic client > rendering & text > renders static text 1`] = `
<div>
  <div>
    Hello World
  </div>
  
</div>
`;

exports[`basic client > rendering & text > should handle lexical scopes correctly 1`] = `
<div>
  <section>
    Nested scope variable
  </section>
  
</div>
`;

exports[`basic client > text rendering > basic operations 1`] = `
<div>
  <div>
    0
  </div>
  <div>
    2
  </div>
  <div>
    5
  </div>
  <div>
    2
  </div>
  
</div>
`;

exports[`basic client > text rendering > renders semi-dynamic text 1`] = `
<div>
  <div>
    Hello World
  </div>
  
</div>
`;

exports[`basic client > text rendering > renders simple JS expression logic correctly 1`] = `
<div>
  <div>
    {"0":"Test"}
  </div>
  <div>
    1
  </div>
  
</div>
`;

exports[`basic client > text rendering > renders static text 1`] = `
<div>
  <div>
    Hello World
  </div>
  
</div>
`;
